### ISSUE TITLE
Boundary conditions in staggered grids

### ISSUE THREAD

**Issue by dschwoerer** [boutproject/BOUT-dev#411]
It seems some Boundary conditions change the values within the domain.

Am I wrong in the assumption that the BC should only change the values outside the domain?

I tested with `test-minmax` and set a few boundary conditions. With `none` the test succeeds, but it fails with:
* `neumann`
* `neumann(1)`
* `dirichlet(1)`
* `neumann_o2`

**Comment by bendudson** [comment]
Correct, it should only change values outside the domain

**Comment by d7919** [comment]
Is this in rc4 or master (note `neumann` and `neumann_o2` are the same in rc4 I think but different in master)?  Is this with 2D and/or 3D fields? What is the location you're using for the staggered grids?

**Comment by d7919** [comment]
Looking at [this line](https://github.com/boutproject/BOUT-dev/blob/v4.0.0-RC/src/mesh/boundary_standard.cxx#L2089) it looks like the boundary operator is offsetting the index in the field being set when dealing with inner boundaries. I'm not very familiar with the staggering but I'm not sure that this is correct -- I'd have thought we should still set just the indices given by the boundary iterator.

**Comment by bendudson** [comment]
If a field is shifted to XLOW, then (assuming MXG=2) two points are outside
the boundary, and the third point is on the boundary. If it were cell
centred then this point would be in the domain, but as it's staggered it's
on the boundary.

**Comment by d7919** [comment]
Ah ok thanks, does this mean that `RGN_NOBNDRY` should skip this point (and if so does it)?

**Comment by bendudson** [comment]
Yes, it probably should skip that. At the moment Field3D::region doesn't
take account of staggered fields. Should be a straightforward fix to check
cell location there.

**Comment by d7919** [comment]
OK, so this is probably the reason that `test-minmax` is impacted -- `min` and `max` use a region iterator to try to avoid the boundaries but if fields are staggered then this will include one of the boundary points.

**Comment by dschwoerer** [comment]
I think it may be a bit more complicated. The staggered grids have a different number of guards cells:
```
1:   .   .   x   x   .   .
2: .   .   x   x   .   .
3: .   .   v   x   v   .
```
1: is the non staggered case.
2: The current interpretation
3: The way it currently is
where `x` is in the domain, `.` is a boundary cell and `v` is on the boundary. Note that the number of guard cells is not symmetric in the staggered case. I am not sure right now whether this means there should be one more guard cell, or whether we have a spare one ...
In case we want to set the value at the boundary `v` should be changed, however if we want to set the derivative, `v` shouldn't be changed, but `.`. In case of 1 guard cell, there would be no cell outside, and we cannot set the derivative ...

**Comment by dschwoerer** [comment]
I think in terms of taking derivatives, if the derivative is taken from the staggerd to the non staggered, the boundary cells are sufficient. However, if we want to take the derivative and have that derivative on the staggered, we need more boundary cells.

Further if the `C2` scheeme is used, it is not really possible to set the derivative at the boundary, if we calculate from stag->nonstag - not sure whether this is a bug, or the way it should be ...
